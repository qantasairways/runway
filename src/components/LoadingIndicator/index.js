import { css, keyframes } from '@emotion/core';
import React from 'react';
import PropTypes from 'prop-types';
import { colours } from '../../theme/airways';

const BUBBLE_SIZE = 1.5;
const DELAY_RANGE = 0.32;

const threeBounce = keyframes`
  from { transform: scale(0); }
  40% { transform: scale(1); }
  80% { transform: scale(0); }
  to { transform: scale(0); }
`;

const loadingIndicator = css`
  position: relative;
  margin: auto;
  text-align: center;
  width: ${BUBBLE_SIZE * 2}rem;
`;

const bounceAnimation = css`
  animation: ${threeBounce} 1.4s ease-in-out infinite both;
  background-color: ${colours.darkGrey};
  border-radius: 100%;
  display: inline-block;
  height: ${(BUBBLE_SIZE / 2).toFixed(2)}rem;
  width: ${(BUBBLE_SIZE / 2).toFixed(2)}rem;
`;

const bounceAnimation1 = css`
  ${bounceAnimation};
  animation-delay: ${DELAY_RANGE}s;
  margin-right: 0.25rem;
`;

const bounceAnimation2 = css`
  ${bounceAnimation};
  animation-delay: ${(DELAY_RANGE / 2).toFixed(2)}s;
  margin-right: 0.25rem;
`;

const screenReaderOnly = css`
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
`;

const LoadingIndicator = ({ screenReaderText }) => (
  <div css={loadingIndicator}>
    <span css={screenReaderOnly}>{screenReaderText}</span>
    <div css={bounceAnimation1} />
    <div css={bounceAnimation2} />
    <div css={bounceAnimation} />
  </div>
);

LoadingIndicator.propTypes = {
  /** Mandatory string to indicate loading for accessibility */
  screenReaderText: PropTypes.string.isRequired
};

export default LoadingIndicator;
