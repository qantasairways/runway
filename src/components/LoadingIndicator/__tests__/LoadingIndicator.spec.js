import React from 'react';
import LoadingIndicator from '..';
import { shallow } from 'enzyme';

describe('LoadingIndicator', () => {
  it('should render correctly with props', () => {
    const component = shallow(<LoadingIndicator screenReaderText="loading" />);

    expect(component).toMatchSnapshot();
  });
});
