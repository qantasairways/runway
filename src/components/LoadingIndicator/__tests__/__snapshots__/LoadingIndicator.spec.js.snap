// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator should render correctly with props 1`] = `
.emotion-4 {
  position: relative;
  margin: auto;
  text-align: center;
  width: 3rem;
}

.emotion-3 {
  -webkit-animation: animation-1oih9nn 1.4s ease-in-out infinite both;
  animation: animation-1oih9nn 1.4s ease-in-out infinite both;
  background-color: #626262;
  border-radius: 100%;
  display: inline-block;
  height: 0.75rem;
  width: 0.75rem;
}

.emotion-1 {
  -webkit-animation: animation-1oih9nn 1.4s ease-in-out infinite both;
  animation: animation-1oih9nn 1.4s ease-in-out infinite both;
  background-color: #626262;
  border-radius: 100%;
  display: inline-block;
  height: 0.75rem;
  width: 0.75rem;
  -webkit-animation-delay: 0.32s;
  animation-delay: 0.32s;
  margin-right: 0.25rem;
}

.emotion-2 {
  -webkit-animation: animation-1oih9nn 1.4s ease-in-out infinite both;
  animation: animation-1oih9nn 1.4s ease-in-out infinite both;
  background-color: #626262;
  border-radius: 100%;
  display: inline-block;
  height: 0.75rem;
  width: 0.75rem;
  -webkit-animation-delay: 0.16s;
  animation-delay: 0.16s;
  margin-right: 0.25rem;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

<div
  className="emotion-4"
>
  <span
    className="emotion-0"
  >
    loading
  </span>
  <div
    className="emotion-1"
  />
  <div
    className="emotion-2"
  />
  <div
    className="emotion-3"
  />
</div>
`;
